File: tests/csv.ml
open Printf
type user = { 
           mutable id: int;
           mutable nom: string; 
           mutable date_naissance: string;
           mutable mail: string;
           mutable telephone: string;
            }

(************************************************************************************************************)
(******************************************    fonction utile    *************************************)
(************************************************************************************************************)

let split chaine separateur =
  let result = ref [] in
  begin
        let ch = ref "" in 
        for i = 0 to String.length chaine - 1 do
               begin
                  if chaine.[i] = separateur then 
                  begin
                      result :=  !ch :: !result; ch := ""
                  end
                  else 
                      ch := !ch^(String.make 1 chaine.[i])
               end
        
        done;
        result :=  !ch :: !result;
  end;
 !result

let lire_file file =
  let entree = (open_in file)
  and result = ref [] in 
  begin
        try
            let ligne = ref (split (input_line entree) ';' ) in
            let nb_champs = (List.length !ligne) in
            while true do
                      if (List.length !ligne) != nb_champs then
                         raise End_of_file 
                      else
                          result := List.append !result [ !ligne ];
                          ligne := (split (input_line entree) ';' ) 
                      done
        with
          | End_of_file -> close_in entree 
  end;
  
  !result




let buffer_file file =
  let entree = (open_in file)
  and result = ref "" in 
  begin
        try
            let ligne = ref (input_line entree)in
            while true do
                      if !ligne != "" then 
                      begin
                          result := !result ^ !ligne ;
                          ligne :=  (input_line entree) 
                      end
                    else raise Exit
                      done
        with
          | End_of_file -> close_in entree 
          | Exit -> print_endline "Fin"
  end;
  !result

(************************************************************************************************************)
(******************************************    fonction pour L'ajout    *************************************)
(************************************************************************************************************)

let write usr =
  let id = (string_of_int (usr.id)) in
  let file = "Inscription.csv" in 
   let x = id^";"^usr.nom^";"^usr.date_naissance^";"^usr.mail^";"^usr.telephone^"\n" in
     let buff = ref (buffer_file file) in 
      let out_chanel = open_out file in 
          buff := !buff^x;
          output_string out_chanel !buff;
          close_out out_chanel
        

 
let write_for_delete usr =
  let id = (string_of_int (usr.id)) in
  let file = "Inscription.csv" in 
   let x = id^";"^usr.nom^";"^usr.date_naissance^";"^usr.mail^";"^usr.telephone in
     let buff = ref (buffer_file file) in 
      let out_chanel = open_out file in 
          buff := !buff^x;
          output_string out_chanel !buff;
          close_out out_chanel
        

 

(************************************************************************************************************)
(******************************************    fonction pour supprimer    ***********************************)
(************************************************************************************************************)


 let clean file =
  let out_chanel = open_out file in 
      output_string out_chanel "";
      close_out out_chanel



  let delete id =
  let file = "Inscription.csv" in 
  let id = string_of_int id in
    let list_ligne = lire_file file in
    let l = ref list_ligne in
    clean file;
      for i = 0 to List.length !l -1 do
      begin 
            let subList = (List.nth !l i) in
            if((List.nth subList 4) <> id) then
                begin
                  let usr_search = ref { 
                                id = int_of_string (List.nth subList 4) ;
                                nom = (List.nth subList 3); 
                                date_naissance = (List.nth subList 2);
                                mail = (List.nth subList 1);
                                telephone = (List.nth subList 0);
                                } in
                 write_for_delete (!usr_search)
                end
          else ()
           
      end
    done

(************************************************************************************************************)
(******************************************    fonction pour la recherche     *******************************)
(************************************************************************************************************)

let search  id_usr = 
     let usr_search = ref { 
                        id = -1 ;
                        nom = ""; 
                        date_naissance = "";
                        mail = "";
                        telephone = "";
                        } in
    let id = (string_of_int (id_usr)) in
    let file = "Inscription.csv" in 
    let list_ligne = lire_file file in
    let l = ref list_ligne in
      for i = 0 to List.length !l -1 do
      begin
        let subList=(List.nth !l i) in
        if((List.nth subList 4) = id) then
        begin
            !usr_search.id <- id_usr;
            !usr_search.nom <- (List.nth subList 3);
            !usr_search.date_naissance <- (List.nth subList 2);
            !usr_search.mail <- (List.nth subList 1);
            !usr_search.telephone <- (List.nth subList 0)
            
        end
      end
        
      done;
  !usr_search 

(************************************************************************************************************)
(******************************************    fonction pour récupérer l'ensemble des users   ***************)
(************************************************************************************************************)

let get_list () = 
  let list_users = ref [] in 
  let file = "Inscription.csv" in 
  let list_ligne = lire_file file in
  let l = ref list_ligne in 
  for i = 0 to List.length !l -1 do
  begin
    let usr_search = ref { 
                                id = int_of_string (List.nth (List.nth !l i) 4) ;
                                nom = (List.nth (List.nth !l i) 3); 
                                date_naissance = (List.nth (List.nth !l i) 2);
                                mail = (List.nth (List.nth !l i) 1);
                                telephone = (List.nth (List.nth !l i) 0);
                                } in
    list_users := usr_search  :: !list_users ;
  end
  done;
  !list_users

File: tests/csv.php
<?php


$usr1 = 
 	[ 
	 	 'id' => 1,
 	 	 'nom' => "ghersa",
 	 	 'date_naissance' => "01/04/1993",
 	 	 'mail' => "m.ghersa.s@gmail.com",
 	 	 'telephone' => "000  000 000 000"
  	] ;
$usr2 = 
 	[ 
	 	 'id' => 2,
 	 	 'nom' => "kasdi",
 	 	 'date_naissance' => "01/03/1994",
 	 	 'mail' => "m.kasdi.h@gmail.com",
 	 	 'telephone' => "111 111 111 111"
  	] ;
$usr3 = 
 	[ 
	 	 'id' => 3,
 	 	 'nom' => "toto",
 	 	 'date_naissance' => "00/11/2016",
 	 	 'mail' => "toto.tata@gmail.com",
 	 	 'telephone' => "333 333 333 333"
  	] ;
$s = ("");
$sNew = ("");
$buf = (array());
$list_users = (array());

	 function split (  $chaine, $separateur ){
  $result = (array());
$ch = ("");
for ($i=0; $i <= (strlen($chaine )-1); $i++) 
 	 { 
 	 		if ( ( ($chaine [$i ])  == $separateur ) ) {
 ($ch )($result );
  ( $ch  = "" );
 
 }	else{
  ( $ch  = (($ch ). ( ($chaine [$i ]) ))  );
 
 }
	 }
 array_unshift ( ($result ) , ($ch ) ) ;
 ($result )
 }



	 function lire_file (  $file ){
  $entree = fopen($file  , "r" ));
$result = (array());

 	 try{ 
 $ligne = (split(input_line($entree ),';');
;
$nb_champs = strlen(($ligne ));
while (true  )
 	 {
 	 		if ( (strlen(($ligne ))!=$nb_champs ) ) {
 echo   ( "Erreur de lecture"." \n " ); 

 }
	else{
  ( $result  = array_merge(($result ),($ligne )) );
 
 }
 ( $ligne  = split(input_line($entree ),';');
 );
 
 	 }

 	 fclose($entree );

 catch (Exception $e){} 
 ($result )
 }



	 function split (  $chaine, $separateur ){
  $result = (array());
$ch = ("");
for ($i=0; $i <= (strlen($chaine )-1); $i++) 
 	 { 
 	 		if ( ( ($chaine [$i ])  == $separateur ) ) {
  array_unshift ( ($result ) , ($ch ) ) ;
  ( $ch  = "" );
 
 }	else{
  ( $ch  = (($ch ). ( ($chaine [$i ]) ))  );
 
 }
	 }
 array_unshift ( ($result ) , ($ch ) ) ;
 ($result )
 }



	 function generate_str (  $lst ){
   
 	 if (sizeof($lst ) == 0) {
 	
 	 }else{

 	 if (sizeof($lst ) >= 1) {
 	for ($i=(strlen($hd )-1); $i >= 0; $i--) 
 	 { 
 	 		if ( ($i  == 0) ) {
  ( $s  = (($s ).($hd [$i ]."\n") )  );
 
 }
	else{
  ( $s  = (($s ).($hd [$i ].";") )  );
 
 }
	 }
 array_unshift ( ($buf ) , ($s ) ) ;
  ( $s  = "" );
 	generate_str ($rst)
 	 }else{

 	 	 }
 
 }



	 function add (  $tab ){
  for ($i=0; $i <= (strlen($tab );
-1); $i++) 
 	 { 
 	 		if ( ($i  == (strlen($tab )-1)) ) {
  ( $sNew  = (($sNew ).($tab [$i ]."\n") )  );
 
 }
	else{
  ( $sNew  = (($sNew ).($tab [$i ].";") )  );
 
 }
	 }

 }



	 function write (  $usr ){
  $id = strval($usr ['id']);
  $file = "Inscription.csv";
$list_ligne = 	lire_file ($file);
$l = ($list_ligne );
	generate_str (($l));
$x = ($usr ['telephone'].(";".($usr ['mail'].(";".($usr ['date_naissance'].(";".($usr ['nom'].(";".$id ) ) ) ) ) ) ) ) ;
$u = split($x ,';');
	add ($u);
 array_unshift ( ($buf ) , ($sNew ) ) ;
 $out_chanel = fopen($file  , "w" ));
for ($i=0; $i <= (strlen(($buf ))-1); $i++) 
 	 { 
 	 	
fputs($out_chanel ,($buf )[$i ]);
	 }
$buf  = array ();
 fclose($out_chanel );
 }



	 function generate_dlt (  $lst, $id ){
   
 	 if (sizeof($lst ) == 0) {
 	 ( $s  = (($s )."")  );
 
 	 }else{

 	 if (sizeof($lst ) >= 1) {
 	for ($i=(strlen($hd )-1); $i >= 0; $i--) 
 	 { 
 	 		if ( ($hd [4] == $id ) ) {
  return ();
 }	else{
 	if ( ($i  == 0) ) {
  ( $s  = (($s ).($hd [$i ]."\n") )  );
 
 }
	else{
  ( $s  = (($s ).($hd [$i ].";") )  );
 
 }

 }

	 }
 array_unshift ( ($buf ) , ($s ) ) ;
  ( $s  = "" );
 generate_dlt($rst ,$id );

 	 }else{

 	 	 }
 
 }



	 function delete (  $id ){
  $id_usr = strval($id );
  $file = "Inscription.csv";
$list_ligne = 	lire_file ($file);
$l = ($list_ligne );
generate_dlt(($l ),$id_usr );
 array_unshift ( ($buf ) , ($sNew ) ) ;
 $out_chanel = fopen($file  , "w" ));
for ($i=(strlen(($buf ))-1); $i >= 0; $i--) 
 	 { 
 	 	
fputs($out_chanel ,($buf )[$i ]);
	 }
$buf  = array ();
 fclose($out_chanel );
 }



	 function generate_search (  $tab, $id, $usr_search ){
   
 	 if (sizeof($tab ) == 0) {
 	()
 	 }else{

 	 if (sizeof($tab ) >= 1) {
 		if ( ($hd [4] == $id ) ) {
 
 ($usr_search )['nom']=$hd [3]; 

 ($usr_search )['date_naissance']=$hd [2]; 

 ($usr_search )['mail']=$hd [1]; 

 ($usr_search )['telephone']=$hd [0]; 

 }	else{
 generate_search($rst ,$id ,$usr_search );

 	 }else{

 	 	 }
 
 }



	 function search (  $id_usr ){
  $usr_search = (
 	[ 
	 	 'id' => $id_usr ,
 	 	 'nom' => "",
 	 	 'date_naissance' => "",
 	 	 'mail' => "",
 	 	 'telephone' => ""
  	] ;
$id = strval($id_usr );
  $file = "Inscription.csv";
$list_ligne = 	lire_file ($file);
$l = ($list_ligne );
generate_search(($l ),$id ,$usr_search );
($usr_search )
 }



	 function generate_lst (  $lst ){
   
 	 if (sizeof($lst ) == 0) {
 	()
 	 }else{

 	 if (sizeof($lst ) >= 1) {
 	$tmp = 
 	[ 
	 	 'id' => intval($hd [4]),
 	 	 'nom' => $hd [3],
 	 	 'date_naissance' => $hd [2],
 	 	 'mail' => $hd [1],
 	 	 'telephone' => $hd [0]
  	] ;
 array_unshift ( ($list_users ) , $tmp  ) ;
 	generate_lst ($rst)
 	 }else{

 	 	 }
 
 }



	 function list (  ){
    $file = "Inscription.csv";
$list_ligne = 	lire_file ($file);
$l = ($list_ligne );
	generate_lst (($l));
($list_users )
 }



?>
